pkg/flow/flowtest/flowtest.go

```diff
@@ -0,0 +1,195 @@
+// Package flowtest provides a script-based testing framework for Flow.
+//
+// [TestScript] accepts a path to a txtar file, where the comment of the txtar
+// file denotes the script to run. Files in the txtar archive are unpacked to a
+// temporary directory as part of the test.
+//
+// See [rsc.io/script] for more information about the scripting language and
+// available default commands.
+//
+// # Flow commands
+//
+// In addition to the default commands provided by [rsc.io/script], the
+// following Flow-specific commands are available:
+//
+//   - `controller.load [file]`: Load a file into the Flow controller.
+//   - `controller.start`: Start the Flow controller.
+//   - `assert.health [component] [health]`: Assert the health of a specific component.
+//
+// Note that `controller.start` should almost always be run in the background
+// by appending a & to the end of the command, otherwise the controller will be
+// terminated immediately after the command exits.
+//
+// Custom commands can be provided by passing [WithExtraCommands] to
+// [TestScript].
+//
+// # Example script
+//
+// The following is an example script which uses local.file:
+//
+//	# This file performs a basic test of loading a component and asserting its
+//	# health.
+//
+//	controller.load main.river
+//	controller.start &
+//
+//	assert.health local.file.example healthy
+//
+//	-- main.river --
+//	local.file "example" {
+//	  filename = "./hello.txt"
+//	}
+//
+//	-- hello.txt --
+//	Hello, world!
+//
+// See the testdata directory for more basic examples.
+package flowtest
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/grafana/agent/pkg/flow"
+	"github.com/grafana/agent/pkg/flow/logging"
+	"github.com/grafana/agent/pkg/flow/tracing"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/tools/txtar"
+	"rsc.io/script"
+)
+
+// TestScript loads the file at filename and executes it as a script. A
+// temporary working directory is created for running the script, and the
+// working directory of the process is changed to the temporary directory for
+// the duration of the call.
+func TestScript(filename string, opts ...TestOption) error {
+	var o options
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	// Create a temporary directory for the scope of our test to operate in.
+	tmpDir, err := os.MkdirTemp("", "flowtest-*")
+	if err != nil {
+		return fmt.Errorf("creating temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	archive, err := txtar.ParseFile(filename)
+	if err != nil {
+		return fmt.Errorf("loading script: %w", err)
+	}
+
+	// Create a test controller for the script to interact with. The test
+	// controller's logs are buffered and printed to stderr on exit. This
+	// prevents log mangling when both the script engine and the Flow controller
```

Comment:
We could put prefixes in logs and print at the same time? Splitting them becomes a problem when we need to correlate the logs of script with the logs of the controller to see what happened in what order.